zuul-core/src/main/java/com/netflix/zuul/netty/filter/BaseZuulFilterRunner.java

```diff
@@ -233,36 +232,34 @@ protected final O filter(final ZuulFilter<I, O> filter, final I inMesg) {
                 return (outMesg != null) ? outMesg : filter.getDefaultOutput(inMesg);
             }
 
-            // async filter
-            PerfMark.startTask(filter.filterName(), "applyAsync");
+            Promise<O> promise;
+            PerfMark.startTask(filter.filterName(), "filterAsync");
             try {
-                final Link nettyToSchedulerLink = PerfMark.linkOut();
+                EventExecutor exec = getChannelHandlerContext(inMesg).executor();
+                promise = new TracingPromise<>(exec, filter.filterName());
+                decrementConcurrency = true;
                 filter.incrementConcurrency();
-                resumer = new FilterChainResumer(inMesg, filter, snapshot, startTime);
-                filter.applyAsync(inMesg)
-                        .doOnSubscribe(() -> {
-                            PerfMark.startTask(filter.filterName(), "onSubscribeAsync");
-                            try {
-                                PerfMark.linkIn(nettyToSchedulerLink);
-                            } finally {
-                              PerfMark.stopTask(filter.filterName(), "onSubscribeAsync");
-                            }
-                        })
-                        .doOnNext(resumer.onNextStarted(nettyToSchedulerLink))
-                        .doOnError(resumer.onErrorStarted(nettyToSchedulerLink))
-                        .doOnCompleted(resumer.onCompletedStarted(nettyToSchedulerLink))
-                        .observeOn(Schedulers.from(getChannelHandlerContext(inMesg).executor()))
-                        .doOnUnsubscribe(resumer::decrementConcurrency)
-                        .subscribe(resumer);
+                filter.filterAsync(promise, inMesg);
             } finally {
-              PerfMark.stopTask(filter.filterName(), "applyAsync");
+                PerfMark.stopTask(filter.filterName(), "filterAsync");
             }
 
+            if (promise.isSuccess()) {
```

Comment:
What's the benefit of handling success here, as opposed to always using the listener callback, regardless of result?

Reply:
Returning early here allows calling the next filter immediately, rather than yielding the thread and handling another request.  It's an optimization.

Reply:
Thought so. Thanks for adding the comments.